lib/active_merchant/billing/gateways/checkout_v2.rb

```diff
@@ -505,6 +507,69 @@ def add_marketplace_data(post, options)
         end
       end
 
+      def add_level_two_three_data(post, options)
+        post[:processing] ||= {}
+        post[:customer] ||= {}
+
+        # American Express only supports Level 2 data.
+        # Only is required add items info in lvl2 data for amex
+        add_items(post, options)
+        add_level_two_data(post, options)
+        add_level_three_data(post, options)
+        add_shipping_data(post, options)
+      end
+
+      def add_items(post, options)
+        items = build_items(options[:line_items])
+        post[:items] = items unless items.nil?
+      end
+
+      def add_level_two_data(post, options)
+        post[:customer][:tax_number] = options[:tax_number] # field no require for amex
+
+        post[:processing].merge!(
+          {
+            order_id: options.dig(:processing, :order_id),
+            tax_amount: options[:tax_amount]
+          }.compact
+        )
+      end
+
+      def add_level_three_data(post, options)
+        post[:processing].merge!(
+          {
+            discount_amount: options[:discount_amount],
+            shipping_amount: options[:shipping_amount],
+            duty_amount: options[:duty_amount]
+          }.compact
+        )
+      end
+
+      def add_shipping_data(post, options)
+        post[:shipping] ||= {}
+        post[:shipping][:from_address_zip] = options[:from_address_zip]
+      end
+
+      def build_items(line_items)
+        return unless line_items
+
+        line_items.map do |item|
+          {
+            # for lvl 2 amex and lvl3 visa/master
```

Comment:
💬 Comment:

two things here to comment:

1) the method has comments that let me know that seems to be the case that there are some params that are to visa/master for lvl3 and some for lvl2 amex, but there is not actually any logic that represents that, no matter the card brand all the fields on this hash are going to be included.

2) seems that perhaps there is a misaligned value from the quantity line to the bottom.

Reply:
1. All of them works for Visa and Mastercard, but only the 3 first are exclusively for amex. This is something also we need to put in documentation, it's risky to made validations for this in the code if for some reason this change in the future we will maintaining this. Also this seams to be something the client should be aware.

If we add any brand validation also we will need to pass the brand in the authorization result for capture cases, according the documentation the level 2 and level 3 values need to be passed in capture.